### ISSUE TITLE
Allow admin curators to add custom strains to the Track database

### ISSUE THREAD

**Issue by jseager7** [pombase/canto#2324]
Currently there's no way for Canto to automatically add a custom strain (entered by the user) to the list of strains shown in the strain picker, meaning that a custom strain can't be re-used across curation sessions (and must be manually entered each time).

The best we can do at the moment is to add the custom strain name as a new strain in the master strain list, but that won't automatically update the strain name in existing curation sessions, meaning the only way to update the strain is to manually edit every genotype in the session (!)

It would really help to have an admin-only option to promote a custom strain to the status of a 'reviewed' strain that can be selected in any session.

(Technically speaking, Canto would add the custom strain name to the Track database, then replace its name in the Curs database with a reference to the strain in the Track database.)

### Implementation

This could involve adding another page to the admin interface, or the features could be built into the strain selection page. Either way, I think at a minimum we'd need to provide the following features:

* **editing the strain name**, in case the admin decides on different naming or formatting; and
* **adding synonyms**, since curators currently have no way to add synonyms when adding custom strains.

In future, we might also need the following features:

* **specifying an abbreviation** for the strain, since it seems pretty likely that we'll be abbreviating strain names in the user interface in future (see https://github.com/pombase/canto/issues/2302);
* **querying for all custom strains**, so that curators can track down sessions with custom strains that might have been missed; and
* **checking for close matches to an existing strain**, in case the user has added a custom strain that is equivalent to a reviewed strain (in this case, the better solution would be to globally edit the custom strain to match the reviewed strain).

### Other concerns

* When a custom strain is promoted to a reviewed strain, it may be necessary to find and replace that custom strain name in all other curation sessions (maybe restricting by taxon ID, just to be safe), since curators may have added the same custom strain in multiple sessions independently.

* Using Canto to add custom strains to the Track database means that the external master copy of the strain list (the one that we load) will go out of sync with Canto. This could make it difficult to keep track of all the strains that are available. I'm not sure of the best way to fix this, but it can probably be solved by periodically exporting the strain list from Canto.

**Comment by jseager7** [comment]
In terms of how this feature might work on the organism and strain page, I think the most sensible way is to have the option appear when a custom strain name is clicked (and only shown when the user is logged in as an admin).

We could have this option shown in a simple drop-down menu, meaning we can also include additional actions (such as 'Delete') that could help solve other issues related to strain management. Here's a mockup of how this might look:

![image](https://user-images.githubusercontent.com/37659591/88285600-4d9f0400-cce7-11ea-8937-8991424029ff.png)

**Comment by jseager7** [comment]
Following further discussion, we've decided it would be better to stick with convention and have the strain list maintained outside of Canto.

With this solution, `canto-load.pl` would need additional logic. I suggest checking whether any new strain (in the list being loaded) matches a custom strain in any curation session. Where there's a match, the custom strain should be replaced with a reference to the new strain that's just been loaded into the Track database.

**Comment by jseager7** [comment]
We might also have cases where the custom strain needs to be merged into an existing strain, as a synonym.

We could do this automatically, by choosing to match against synonyms (as well as primary names) when searching for matching strains in curation sessions (following the process above), but this might make the matching too broad.

Alternatively, we could add a sub-option to `canto_load.pl --merge-strains` that keeps the merged strain name as a synonym of the 'winning' strain name, instead of discarding the old strain name. This means a bit more manual work, since the custom strain has to be added to Track (using the process [above](https://github.com/pombase/canto/issues/2324#issuecomment-667985476)) and then immediately merged.

**Comment by jseager7** [comment]
@kimrutherford Could you focus on this when you have time?

Now that the new strain list is loaded, I'd like to work through our custom strains in PHI-Canto to see which ones should be promoted to the master list.

In order to do this, I think I need `canto-load.pl` (or some other script) to perform the following actions:

1. For each newly-added strain in the loaded strain list, check if the strain name exists as a custom name in any of the Curs databases (a 'newly-added' strain is a strain whose name is not already in the Track `strain` table).

2. Where a custom strain exists, its name should be replaced with a reference to the corresponding strain in the Track database.

Basically, the plan is to promote custom strains to the master list by simply adding them to the master list, and trusting the loading code to handle the rest automatically.

If you can think of a better way to do this, I'd be happy to hear it.

**Comment by kimrutherford** [comment]
Hi James.  Sorry for the slow reply.

I think it's a good plan.  Can you think of any case where we wouldn't want to promote a strain to the main track database?  If not, I'll go ahead and implement this.

**Comment by jseager7** [comment]
> Can you think of any case where we wouldn't want to promote a strain to the main track database?

Do you mean when we'd want to prevent the strain being promoted? I think the only cases where we'd want to forbid this are:

* when the promoted strain name already exists as primary strain name in the Track database; or

* when the promoted strain name already exists as synonym of a strain in the Track database.

Having said that, the second case probably deserves a warning rather than an error, since there may really be cases where two different strains have the same synonym.

More generally, we might have cases where we want to delay promoting a strain because it needs further investigation, but I doubt we need any technical measures for this case.

**Comment by kimrutherford** [comment]
Hi James.  I've re-read all your comments and I understand better what's needed.

> I suggest checking whether any new strain (in the list being loaded) matches a custom strain in any curation session. Where there's a match, the custom strain should be replaced with a reference to the new strain that's just been loaded into the Track database.

I think this can all be automatic except for the case you mention where two strains have the same synonym.  I think that will require manual intervention.  Can we chat about this on Skype next week?  I want to make sure I understand any other special cases.

**Comment by kimrutherford** [comment]
Hi James.  I've modified `canto_load.pl --strains` so that the sessions are fixed up at the same time, as we discussed.  The new strains will be fixed to point to the main track database when possible.  I've tested it locally and added a Canto test case for the code.  Could you have a go when you get a chance?

There is now an error if two strains of an organism have the same synonym.  Should that be a warning rather than an error?

**Comment by jseager7** [comment]
> There is now an error if two strains of an organism have the same synonym. Should that be a warning rather than an error?

My gut feeling is that it should be a warning, especially since there are examples of non-unique synonyms in established databases like MGI (see image).

![image](https://user-images.githubusercontent.com/37659591/94435662-57376300-0193-11eb-886c-4e112c66325b.png)


Even in the case where a synonym is added by mistake (an error case), I think it's fairly simple to amend the synonyms of a strain to fix the mistake.

**Comment by jseager7** [comment]
> My gut feeling is that it should be a warning

Now that I've looked into this, it definitely should be a warning (not an error), since the new PHI-base strain list is already using at least two of these previously mentioned mouse strains, and they have shared synonyms:

strain | synonyms
-- | --
C57BL/6 | B6, Black 6, C57BL/6-Mx1<s1>
C57BL/6J | B6, B6J, Black 6, C57 Black

I then checked PHI-Canto, where there are even more suggestions:

![image](https://user-images.githubusercontent.com/37659591/94436335-3c192300-0194-11eb-80fb-36018556db68.png)

**Comment by jseager7** [comment]
And to confirm, this prevents me from testing the feature with our strain list, since both of the lists have duplicate synonyms. Here's an example from the pathogen strain list.

> load failed - strain synonym "H99" for strain "var grubii serotype A strain H99" for taxon 5207 is duplicated in /import_export/pathogen_strains.csv

**Comment by kimrutherford** [comment]
Hi James.

I've had a go at turning the error into warnings.  It warns about the case where two strains have the same synonym and then warns if a session strain has the same name as the synonyms.  And it also warns about the case where there is ambiguity because a session strain matches a strain name and a synonym.

Examples:

I've added A1160test as a strain name in my session.  Then I added this to my strains file:
```csv
746128,Aspergillus fumigatus,A1160C,"A1160test"
746128,Aspergillus fumigatus,A1160P,"A1160test"
```

It gives this output:
```
Warning: strain synonym "A1160test" for strain "A1160P" for taxon 746128 is duplicated in t/data/pathogen_strains_sample.csv
The strain "746128:A1160test" in session 0d2bc8d23a8667f4 is ambiguous as there is more than one strain in the track DB with that name as a synonym.
Strains of taxon 746128 with "A1160test" as a synonym: A1160C, A1160P
```

Example of the second case:

With this in the input file:
```csv
746128,Aspergillus fumigatus,A1160C,"A1160test"
746128,Aspergillus fumigatus,A1160test
```
I get this output:
```
The strain "746128:A1160test" in session 0d2bc8d23a8667f4 is ambiguous as there is a strain in the track database with that name and also a strain synonym with that name.
Strains with "A1160test" as a synonym: A1160C
```

I've checked in this change.  Could you give it a try?

**Comment by kimrutherford** [comment]
I've just tweaked this message:
```
Strains with "A1160test" as a synonym: A1160C
```
to be:
```
Strains of taxon 746128 with "A1160test" as a synonym: A1160C
```

**Comment by jseager7** [comment]
Thanks for the changes. I'll try to test them today. 

Just to clarify, what actually happens when the strain name in the session has an ambiguous mapping to the strain file (because the session strain's name matches the synonym of two or more strains in the strain file)? Does the loader skip the ambiguous strains? Does it load the strains but skip promoting the session strains? Or does it arbitrarily choose one of the new strains as the 'winning' strain that the session strain should be promoted to?

**Comment by kimrutherford** [comment]
> Does it load the strains but skip promoting the session strain?

Yep!  It warns and doesn't change any strains in sessions that match.  Those ones will need manual fixing or and extra option on the script to disambiguate things.

**Comment by jseager7** [comment]
**TL;DR:** I can't find any problems after testing, so feel free to close.

I've tested adding a session strain and then promoting it by loading a matching strain in the strain file, and that works as expected. The new strain lists load correctly without errors (and I get the expected warnings about non-unique synonyms).

I've also tested the ambiguous loading case with some real data from our PHI-base strain list, and that works too. See below for the process.

I can open a new issue about better solutions for handling the case where strains can't be automatically promoted (due to ambiguous synonyms). Currently I don't think we have an easy way to rename strains that only exist in the session, but that can afford to wait until it's needed.

- - -

To test the ambiguous loading, I deleted all the strains in my local copy of Canto, added 'H99' as a session strain (_Cryptococcus neoformans_), then loaded two strains that both have 'H99' as synonyms: 

strain_id|organism_id|strain_name|strain_abbreviation
---|---|---|---
1|87|serotype A strain H99|
2|87|var. grubii serotype A strain H99|

strainsynonym_id|strain_id|synonym
---|---|---
1|1|H99
2|2|H99

I got all of the expected error messages:

`Warning: strain synonym "H99" for strain "var grubii serotype A strain H99" for taxon 5207 is duplicated in /import_export/test_strains.tsv`  
`The strain "5207:H99" in session c7313b53a1533f0c is ambiguous as there is more than one strain in the track DB with that name as a synonym.`  
`Strains of taxon 5207 with "H99" as a synonym: serotype A strain H99, var. grubii serotype A strain H99`

I then checked and confirmed that the 'H99' session strain wasn't promoted.
